tutorials/matchmaker101/director/main.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"math/rand"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"open-match.dev/open-match/pkg/pb"
+)
+
+// The Director in this tutorial does the following:
+// - Generates the profiles that it will poll for.
+// - Polls Open Match backend for those profiles.
+//
+// The Director can be configured to make a single fetch for all the profiles
+// or can keep polling forever at a configured interval.
+
+var (
+	omBackendEndpoint       = "om-backend.open-match.svc.cluster.local:50505"
+	functionHostName        = "mm101-tutorial-matchfunction.mm101-tutorial.svc.cluster.local"
+	functionPort      int32 = 50502
+	interval                = 5000 // Millisecond duration for the polling interval.
+	fetchForever            = true // Fetch once if false, forever if true.
+)
+
+func main() {
+	// Connect to Open Match Backend.
+	conn, err := grpc.Dial(omBackendEndpoint, grpc.WithInsecure())
+	if err != nil {
+		log.Fatalf("Failed to connect to Open Match Backend, got %w", err)
+	}
+
+	defer conn.Close()
+	be := pb.NewBackendClient(conn)
+
+	// Generate the profiles to fetch matches for.
+	profiles := generateProfiles()
+	log.Printf("Fetching matches for %v profiles", len(profiles))
+
+	// Fetch matches for each profile and then wait for all the calls to return
+	// before starting the next iteration.
+	for {
+		var wg sync.WaitGroup
+		for _, p := range profiles {
+			wg.Add(1)
+			go func(wg *sync.WaitGroup, p *pb.MatchProfile) {
+				defer wg.Done()
+				matches, err := fetch(be, p)
+				if err != nil {
+					log.Printf("Failed to fetch matches for profile %v, got %w", p.GetName(), err)
+					return
+				}
+
+				log.Printf("Generated %v matches for profile %v", len(matches), p.GetName())
+
+				// Make random assignment for all the matches returned for this profile.
+				if err := assign(be, matches); err != nil {
+					log.Printf("Failed to assign servers to matches, got %w", err)
+					return
+				}
+			}(&wg, p)
+		}
+
+		wg.Wait()
+		if !fetchForever {
+			break
+		}
+		time.Sleep(time.Millisecond * time.Duration(interval))
```

Comment:
use a time.Ticker, as it will lead to a consistent interval time without varying because of how long the fetch matches and assignments takes.

Reply:
(unless it takes too long, in which case ticker will drop missed intervals.)

Reply:
Basic tutorial - really easy to understand the logs if you follow a fetch-assign-wait pattern. No requirement to tick at consistent interval time. If fetch / assign blocks for whatever reason, I would rather have this loop block and know there is a problem.